libs/ui/src/lib/data-table/SalesforceRecordDataTable.tsx

```diff
@@ -99,6 +99,7 @@ export const SalesforceRecordDataTable: FunctionComponent<SalesforceRecordDataTa
     const [nextRecordsUrl, setNextRecordsUrl] = useState<Maybe<string>>(null);
     const [globalFilter, setGlobalFilter] = useState<string | null>(null);
     const [selectedRows, setSelectedRows] = useState<ReadonlySet<string>>(() => new Set());
+    const [visibleRecordCount, setVisibleRecordCount] = useState<number>(0);
```

Comment:
@joiecosby - usually it is best if you can try to initialize any state values with an accurate value if possible, and I think here we can. Reason being is that it ensures that if the UI renders initially it will show the correct value instead of showing an incorrect value momentarily then showing the correct value.

Also, the type can be inferred by the value so you generally don't need to include a generic type (`<number>`) unless the type cannot be inferred.

```suggestion
    const [visibleRecordCount, setVisibleRecordCount] = useState(records.length);
```

Also, you might notice that many places use the function syntax `useState(() => records.length);` -> this syntax should be used if deriving the value is expensive (anything that is not a simple value being used) because the function syntax gets called lazily, where the normal syntax gets called on every render no matter what even if there is already a value stored in state.

Reply:
@joiecosby - Here is an example that I just came across. State was initialized as empty array then an useEffect was used to set the data, which happens on a subsequent render.

In this case, the computation was sort of expensive (loop through a bunch of items) so in addition to setting it on init with a function (lazy init), I also used a custom hook in the codebase called `useNonInitialEffect` - it just means it does not run on first render, but runs on ever change after that.

I also extracted to a function, outside the component, since there was a lot of code. The reason the function was moved outside the component was because that means it does not have to get re-created on every render. It also means the function must take in all the data as a parameter instead of having direct access to variables within the component's scope.

Here is a demo of the UI experience before and after:
BEFORE: https://jam.dev/c/f8cf3c05-1a10-4e37-b0e7-90887458f8b5
AFTER: https://jam.dev/c/b39fb3fa-7121-45c5-8ace-bea276d3ceee

![image](https://user-images.githubusercontent.com/5461649/234320547-ca82426c-9544-4635-a295-52a33a2b5fbb.png)

Reply:
That is a great visual representation of what you are trying to explain. So because the logic for setting visibleRecordCount is straightforward and not costly, we can set it by just using `useState(records.length);` 

Is the logic below, another good example of what you are trying to get across? That we init at rendering by calling getModalTitle() and then using useNonInitialEffect() to render any changes more forward after the initial init. 

Quick question, does useNonInitialEffect() know to track action by passing in [action] at the end? 

```
  const [modalTitle, setModalTitle] = useState(() => getModalTitle(action));

  const [downloadModalOpen, setDownloadModalOpen] = useState(false);

  useNonInitialEffect(() => {
    setModalTitle(getModalTitle(action));
  }, [action]);
```

Reply:
Two more questions for you:

1. since records is getting set in a const above and is getting set to any during the init do we wouldn't ever have to worry about record every being null, correct?

` const [visibleRecordCount, setVisibleRecordCount] = useState(records?.length);`

2. If records isn't being set until the useEffect is being called, would visibleRecordCount technically still be 0 on it?

Reply:
`useNonInitialEffect` - this is just a custom thing I built (or probably found on a blog article and slightly modified) - `useEffect` runs whenever dependencies change and cause a re-render, with `useNonInitialEffect` is the exact same except won't run on the very first render.
This is sort of like a micro-optimization, but sometimes - mostly in very complex components, it does make a difference.

> do we ever have to worry about setting the const like below
If `records` can be undefined in the props type definition, then we would absolutely need to do `useState(records?.length);`  - which as a result, `visibleRecordCount` would have the type `number | undefined`.
Many times it is nice to ensure a value has a value, in this case `useState(records?.length || 0);` would be a good solution.

> If records isn't being set until the useEffect is being called, would visibleRecordCount technically still be 0 on it?
Yes - BUT there is a caveat. Once records is defined, it shows up in the ui right away THEN the useEffect is called on a second render - so there is still room for a jumpy UI. Usually the jumpy UI only happens on the initial render unless there is a lot of dependent updates being made that result in spanning multiple renders.

You will see at some places around the codebase `useReducer` - this is a solution to that problem above, where instead of having a bunch of `useEffect` all updating state based on prior useEffect, all the updates can be made at once. (useReducer isn't used very often and is only really only helpful in fairly complex situations)